aspnetcore/blazor/javascript-interoperability/import-export-interop.md

```diff
@@ -0,0 +1,417 @@
+
+The `setMessage` method:
+
+* Calls `globalThis.getDotnetRuntime(0)` to expose the WebAssembly .NET runtime instance for calling exported .NET methods.
+* Obtains the app assembly's JS exports. The name of the app's assembly in the following example is `BlazorSample`.
+* Calls the `BlazorSample.Pages.CallDotNet1.GetMessageFromDotnet` method from the exports (`exports`). The returned value, which is the welcome message, is assigned to the `CallDotNet1` component's `<span>` text. The app's namespace is `BlazorSample`, and the `CallDotNet1` component's namespace is `BlazorSample.Pages`.
+
+`Pages/CallDotNet1.razor.js`:
+
+```javascript
+export async function setMessage() {
+  const { getAssemblyExports } = await globalThis.getDotnetRuntime(0);
+  var exports = await getAssemblyExports("BlazorSample.dll");
+
+  document.getElementById("result").innerText = 
+    exports.BlazorSample.Pages.CallDotNet1.GetMessageFromDotnet();
+}
+```
+
+If you need to iteratively make code changes in JS files and force a browser to reload the files (cache busting), we recommend using browser [developer tools](https://developer.mozilla.org/docs/Glossary/Developer_Tools) with static asset caching disabled. For more information, access the documentation for the developer tools associated with your browser:
+
+* [Chrome DevTools](https://developer.chrome.com/docs/devtools/)
+* [Firefox Developer Tools](https://developer.mozilla.org/docs/Tools)
+* [Microsoft Edge Developer Tools overview](/microsoft-edge/devtools-guide-chromium/)
+
+## Use of a single JavaScript module across components
+
+*Before following the guidance in this section, read the [Call JavaScript from .NET](#call-javascript-from-net) and [Call .NET from JavaScript](#call-net-from-javascript) sections of this article, which provide general guidance on .NET JS `[JSImport]`/`[JSExport]` interop.*
+
+The example in this section shows how to use JS interop from a shared JS module. The following components, classes, C# methods, and JS functions are used:
+
+* `Interop` class (`Interop.cs`): Sets up import and export JS interop with the `[JSImport]` and `[JSExport]` attributes for a module named `Interop`.
+  * `GetWelcomeMessage`: .NET method that calls the imported `getMessage` JS function.
+  * `SetWelcomeMessage`: .NET method that calls the imported `setMessage` JS function.
+  * `GetMessageFromDotnet`: An exported C# method that returns a welcome message string when called from JS.
+* `wwwroot/js/interop.js` file: Contains the JS functions.
+  * `getMessage`: Returns a welcome message when called by C# code in a component.
+  * `setMessage`: Calls the `GetMessageFromDotnet` C# method and assigns the returned welcome message to a DOM `<span>` element.
+* `App` component (`App.razor`): Loads the module from `wwwroot/js/interop.js` by calling `JSHost.ImportAsync` in the [`OnInitializedAsync` event](xref:blazor/components/lifecycle#component-initialization-oninitializedasync).
```

Comment:
I believe the module loading can go in `Program.cs`.

Reply:
Yes, that works. Are we going to show both, or is only ever doing it in `Program.cs` the most sensible approach?

Reply:
Another question is where this would be done for a library?

Reply:
> ... and I now guess that he does that because of this var set in the module ...

I mentioned this in my other comment, but I believe this is actually because JS interop won't work in `OnInitializedAsync()` during Blazor Server prerendering (which isn't a concern for this new JS interop API, which is Wasm-only).

> Yes, that works. Are we going to show both, or is only ever doing it in `Program.cs` the most sensible approach?

If you're writing an RCL, `Program.cs` is definitely not an option. I think the best choice in the library case is to place module loading in component initialization (as is done in most of your examples). A possible downside is that could be slightly confusing because it gives the illusion that lifetime of the loaded module is somehow related to lifetime of the component (as it might be for modules imported as an `IJSObjectReference`). It would be cool if there was a better/easier/more automatic way to load JS modules using this new API, but I think that would have to come in the form of an additional framework feature.

But if you're going to place module loading in `App.razor`, you're probably not writing an RCL, so the logic can just go in `Program.cs`. IMO, that's slightly cleaner because the `App` component doesn't really have anything to do with module loading, and `Program.cs` is already used for app-wide configuration.

cc @SteveSandersonMS again in case you have any objections to any of this 😄

Reply:
I'll call out only `Program.cs` for the example. I'll indicate that this particular scenario ... the 3rd section approach ... is best applied to apps, not RCLs.